runtime/commonMain/src/kotlinx/serialization/SerialDescriptorBuilder.kt

```diff
@@ -0,0 +1,240 @@
+    private val uniqueNames: MutableSet<String> = HashSet(elementsCount)
+    internal val elementDescriptors: MutableList<SerialDescriptor> = ArrayList(elementsCount)
+    internal val elementAnnotations: MutableList<List<Annotation>> = ArrayList(elementsCount)
+    internal val elementOptionality: MutableList<Boolean> = ArrayList(elementsCount)
+
+    /**
+     * Add an element with a given [name][elementName], [descriptor],
+     * type annotations and optionality the resulting descriptor.
+     *
+     * Example of usage:
+     * ```
+     * class Data(
+     *     val intField: Int? = null, // Optional, has default value
+     *     @ProtoId(1) val longField: Long
+     * )
+     *
+     * // Corresponding descriptor
+     * SerialDescriptor("package.Data", 2) {
+     *     element<Int?>("intField", isOptional = true)
+     *     element<Long>("longField", annotations = listOf(protoIdAnnotationInstance))
+     * }
+     * ```
+     */
+    public fun element(
+        elementName: String,
+        descriptor: SerialDescriptor,
+        annotations: List<Annotation> = emptyList(),
+        isOptional: Boolean = false
+    ) {
+        checkOverflow(elementName)
+        elementNames += elementName
+        if (!uniqueNames.add(elementName)) {
+            error("Element with name '$elementName' is already registered")
+        }
+        elementDescriptors += descriptor
+        elementAnnotations += annotations
+        elementOptionality += isOptional
+    }
+
+    /**
+     * Reified version of [element] function that
+     * extract descriptor using `serializer<T>().descriptor` call with the all the restrictions of `serializer<T>().descriptor`.
+     */
+    @ImplicitReflectionSerializer
+    public inline fun <reified T> element(
+        elementName: String,
+        annotations: List<Annotation> = emptyList(),
+        isOptional: Boolean = false
+    ) {
+        val descriptor = serializer<T>().descriptor
+        element(elementName, descriptor, annotations, isOptional)
+    }
+
+    internal fun verify() {
+        if (elementNames.size != elementsCount) {
+            error("Expected $elementsCount elements, but has only ${elementNames.size}")
+        }
+    }
+
+    private fun checkOverflow(elementName: String) {
+        if (elementNames.size < elementsCount) return
+        error("Declared only $elementsCount elements, but tried to add one more: $elementName")
+    }
+}
+
+// All these methods are extensions to avoid top-level scope pollution
+
+/**
+ * Retrieves descriptor of type [T] using reified [serializer] function.
+ */
+@ImplicitReflectionSerializer
+public inline fun <reified T> SerialDescriptorBuilder.descriptor(): SerialDescriptor = serializer<T>().descriptor
+
+/**
+ * Creates descriptor for the type `List<T>` where `T` is the type associated with [typeDescriptor].
+ */
+public fun SerialDescriptorBuilder.listDescriptor(typeDescriptor: SerialDescriptor): SerialDescriptor {
```

Comment:
We have convention functions like `.list` to do `IntSerializer.list`, maybe make this API similar?

Reply:
Then we'll lose the consistency and discoverability of reified version (`listDescriptor<Int>`).
And using `IntSerializer.descriptor.list` is not really readable